File: backend/internal/service/worker.py
from __future__ import annotations

from dao.repositories.worker import WorkerRepository
from pkg.apis.worker.models import WorkerHeartbeat, WorkerResource
from pkg.config.settings import Settings
from pkg.runtime.status import WorkerStatus

from internal.exceptions import NotFoundError, WorkerAuthError


class WorkerService:
    def __init__(self, workers: WorkerRepository, settings: Settings) -> None:
        self.workers = workers
        self.settings = settings

    def authenticate_token(self, token: str | None) -> None:
        if token != self.settings.worker_token:
            raise WorkerAuthError()

    async def register(self, resource: WorkerResource) -> WorkerResource:
        resource.status = WorkerStatus.ONLINE.value
        worker = await self.workers.upsert_registration(resource)
        return await self.workers.as_resource(worker)

    async def heartbeat(self, payload: WorkerHeartbeat) -> WorkerResource:
        worker = await self.workers.record_heartbeat(
            payload.worker_id,
            status=payload.status,
            resources=payload.resources,
            models=payload.models,
            running_requests=payload.running_requests,
            metadata=payload.metadata,
        )
        if worker is None:
            raise NotFoundError(f"Worker {payload.worker_id} is not registered")
        return await self.workers.as_resource(worker)

    async def list_workers(self) -> list[WorkerResource]:
        return await self.workers.list_resources()

    async def inspect(self, worker_id: str) -> WorkerResource:
        worker = await self.workers.get_by_worker_id(worker_id)
        if worker is None:
            raise NotFoundError(f"Worker {worker_id} not found")
        return await self.workers.as_resource(worker)

File: backend/internal/service/worker.go
package service

import (
	"context"

	"github.com/spider/spider/dao/store"
	"github.com/spider/spider/internal/spidererrors"
	"github.com/spider/spider/pkg/apis"
)

type WorkerService struct {
	Workers *store.WorkerRepo
	Token   string
}

func (s *WorkerService) AuthenticateToken(token string) error {
	if token != s.Token {
		return spidererrors.WorkerAuth("Invalid worker token")
	}
	return nil
}

func (s *WorkerService) Register(ctx context.Context, resource apis.WorkerResource) (apis.WorkerResource, error) {
	_, err := s.Workers.UpsertRegistration(ctx, resource)
	if err != nil {
		return apis.WorkerResource{}, err
	}
	w, err := s.Workers.GetByWorkerID(ctx, resource.WorkerID)
	if err != nil {
		return apis.WorkerResource{}, err
	}
	return s.Workers.AsResource(ctx, w)
}

func (s *WorkerService) Heartbeat(ctx context.Context, hb apis.WorkerHeartbeat) (apis.WorkerResource, error) {
	w, err := s.Workers.RecordHeartbeat(ctx, hb)
	if err != nil || w == nil {
		return apis.WorkerResource{}, spidererrors.NotFound("Worker not found")
	}
	return s.Workers.AsResource(ctx, w)
}

func (s *WorkerService) ListWorkers(ctx context.Context) ([]apis.WorkerResource, error) {
	return s.Workers.ListResources(ctx)
}

func (s *WorkerService) Inspect(ctx context.Context, workerID string) (apis.WorkerResource, error) {
	w, err := s.Workers.GetByWorkerID(ctx, workerID)
	if err != nil {
		return apis.WorkerResource{}, spidererrors.NotFound("Worker not found")
	}
	return s.Workers.AsResource(ctx, w)
}
